### PR TITLE
Implemented authentication and authorization

### PR DESCRIPTION
**Pull request by Ekman1991**
This PR adds authentication with FaunaDB and GraphQL authorization via their roles system.
It also adds server side rendering with Apollo queries.

The flow is like this. A user is logging in with the JS client of Fauna and upon successful login we are receiving a secret token from Fauna. This token is serialized and added as a cookie for the user. This token is later used when making queries too GraphQL to assure that only authenticated users can query data.

To hide a page behind authentication wrap the page within the HOC `withAuthSync` for example:
`export default withAuthSync(IndexPage);`. This will redirect unauthenticated users to `/login`.

The Apollo lib is no longer wrapping the complete app (to be able to handle SSR). To be able to make GraphQL queries the page needs to use the HOC `withApollo`. The Apollo HOC is reading the secret from a cookie so an example of a page that needs to make GraphQL queries is:
`export default withApollo(withAuthSync(QuestionaireResultsPage), { getDataFromTree });`.

If a user is fiddling with the cookies they are signed out.

For now there's only one role `MANAGER` and the creation of users are handled manually. Although there are logic to create a user via the app no front-end is implemented for signup.

### CODE REVIEW

File: components/Layout/Layout.tsx [hunk 1]
```diff
@@ -24,6 +24,8 @@ import {
   CollapseBtn,
   SidebarTrigger
 } from "@mui-treasury/layout";
+import { logout } from '../../utils/auth'
```
Comment (daniel-primea) [inline components/Layout/Layout.tsx:27]: Använd root based imports '~/utils/auth'

File: package.json [hunk 2]
```diff
@@ -16,6 +16,11 @@
     "@mui-treasury/layout": "latest",
     "@types/node": "latest",
     "@types/react": "latest",
+    "@types/cookie": "^0.3.3",
+    "cookie": "^0.4.0",
+    "dotenv": "8.2.0",
```
Comment (daniel-primea) [inline package.json:21]: Är denna dependency låst vid 8.2.0 av någon speciell anledning?

File: pages/api/login.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { query as q } from 'faunadb'
+import { serverClient, serializeFaunaCookie } from '../../utils/fauna-auth'
```
Comment (daniel-primea) [inline pages/api/login.tsx:2]: Använd root based imports '~/utils/fauna-auth'

File: pages/api/logout.tsx [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import { query as q } from 'faunadb'
+import cookie from 'cookie'
+import { faunaClient, FAUNA_SECRET_COOKIE } from '../../utils/fauna-auth'
```
Comment (daniel-primea) [inline pages/api/logout.tsx:3]: Använd root based imports '~/utils/fauna-auth'

File: pages/api/signup.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
+import { query as q } from 'faunadb'
+import { serverClient, serializeFaunaCookie } from '../../utils/fauna-auth'
```
Comment (daniel-primea) [inline pages/api/signup.tsx:2]: Använd root based imports '~/utils/fauna-auth'

File: pages/api/logout.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
+import { query as q } from 'faunadb'
```
Comment (daniel-primea) [inline pages/api/logout.tsx:1]: Någon speciell anledning till att query får namnet q? Osäker på om det är tillräckligt tydligt namn.

File: pages/login.tsx [hunk 7]
```diff
@@ -183,14 +183,14 @@ const LoginPage: React.FunctionComponent = () => {
...
+                    id="email"
                     formControlProps={{
                       fullWidth: true
                     }}
                     inputProps={{
-                      value: values.username,
-                      onChange: handleChange("username"),
+                      value: values.email,
+                      onChange: handleChange("email"),
                       type: "text",
```
Comment (daniel-primea) [inline pages/login.tsx:194]: Om du ändrar till email så bör type också ändras till "email"

File: utils/auth.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+          maxAge: -1,
+          httpOnly: true,
+          path: '/',
+      })
+      res?.setHeader('Set-Cookie', cookieSerialized)
+      res?.writeHead(302, { Location: '/login' })
+      res?.end()
+    }
+
+    return { user: 'phil' }
```
Comment (daniel-primea) [inline utils/auth.tsx:63]: `return { user: 'phil' }` Ser hårdkodat ut?

File: pages/api/logout.tsx (hunk 6 above)
Comment (daniel-primea) [reply pages/api/logout.tsx:1]: Fast det kanske är standard. Blir ju långa chains annars ser jag.

File: package.json [hunk 9]
```diff
@@ -16,6 +16,11 @@
...
+    "faunadb": "^2.13.1",
+    "js-cookie": "^2.2.0",
```
Comment (daniel-primea) [inline package.json:23]: Kan inte se att `js-cookie` används. Antar att `cookie` räcker?

File: utils/auth.tsx (hunk 8 above)
Comment (Ekman1991) [reply utils/auth.tsx]: Nice catch, hade helt missat det :)

**Comment by daniel-primea** [comment]
#### 📝Changed routes:


| [**/about**](https://360-analys-git-auth.primea-ab.now.sh/about) | [**/**](https://360-analys-git-auth.primea-ab.now.sh/) |
|:-:|:-:|
|<a href="https://360-analys-git-auth.primea-ab.now.sh/about">Screenshot of /about</a><br /><sup><a href="https://images.weserv.nl?url=https://s.microlink.io/ESop9l2h_0djLVXrwPyVQa8WgMA_.png">(view full size)</a> |<a href="https://360-analys-git-auth.primea-ab.now.sh/">Screenshot of /</a><br /><sup><a href="https://images.weserv.nl?url=https://s.microlink.io/KiCEBd2Pn64wZyym4MEQ_nT7kaw_.png">(view full size)</a> |



| [**/login**](https://360-analys-git-auth.primea-ab.now.sh/login) | [**/questionaire/\[id\]**](https://360-analys-git-auth.primea-ab.now.sh/questionaire/[id]) |
|:-:|:-:|
|<a href="https://360-analys-git-auth.primea-ab.now.sh/login">Screenshot of /login</a><br /><sup><a href="https://images.weserv.nl?url=https://s.microlink.io/EZ0UQF_YamChb_R_VIY_jX6Ef7w_.png">(view full size)</a> |<a href="https://360-analys-git-auth.primea-ab.now.sh/questionaire/[id]">Screenshot of /questionaire/[id]</a><br /><sup><a href="https://images.weserv.nl?url=https://s.microlink.io/cnEBrwMgl2kMGWwlF1EMqOz99bk_.png">(view full size)</a> |



| [**…stionaire/\[id\]/results**](https://360-analys-git-auth.primea-ab.now.sh/questionaire/[id]/results) |
|:-:|
|<a href="https://360-analys-git-auth.primea-ab.now.sh/questionaire/[id]/results">Screenshot of /questionaire/[id]/results</a><br /><sup><a href="https://images.weserv.nl?url=https://s.microlink.io/gc90Gt8GICzK5m0u5J6fLCGddHA_.png">(view full size)</a> |



And 4 other routes:
- [**/api/auth-types**](https://360-analys-git-auth.primea-ab.now.sh/api/auth-types)
- [**/api/login**](https://360-analys-git-auth.primea-ab.now.sh/api/login)
- [**/api/logout**](https://360-analys-git-auth.primea-ab.now.sh/api/logout)
- [**/api/signup**](https://360-analys-git-auth.primea-ab.now.sh/api/signup)



Commit 8889822005e07e332b9db9dc381db71d1741891b (https://360-analys-git-auth.primea-ab.now.sh).

File: package.json (hunk 9 above)
Comment (Ekman1991) [reply package.json]: Helt rätt, jag tog bort den och testade och den verkade inte användas alls :)

File: package.json (hunk 2 above)
Comment (Ekman1991) [reply package.json]: Hmm jag tror inte det, nog bara en miss :)

File: pages/api/logout.tsx (hunk 6 above)
Comment (Ekman1991) [reply pages/api/logout.tsx:1]: Tror dom hade så i deras exempel :)

**Comment by daniel-primea** [comment]
Bara o merga :)

**Pull request merged**